File: lib/voice-engine/voice-idiosyncrasy-extractor.ts
// PR Sprint 7.22 Sprint E.2 — E1 voice idiosyncrasies extractor
// (TS port of
// Helm SEO/helm-adaptive-voice-engine/voice_idiosyncrasy_extractor.py).
//
// Extracts statistical voice patterns from a list of past post texts.
// The output is a VoiceIdiosyncrasies object that the prompt builder
// injects as a WRITER VOICE PROFILE block of concrete rules
// (em-dash rate, fragment ratio, common openers, hedging behavior,
// etc.) so the model can match the writer's voice as an explicit
// set of constraints rather than imitating loose samples.
//
// Pure functions — no DB, no network, no async. The
// maybe-refresh-idiosyncrasies helper does the DB query +
// freshness logic and calls into this module.
//
// Canonical TS source-of-truth. Python upstream at
// lib/voice-engine/voice_idiosyncrasy_extractor.py (and
// Helm SEO/helm-adaptive-voice-engine/) MUST stay in lockstep.

import type { VoiceIdiosyncrasies } from '@/lib/types/brand';

// ============================================================
// Tokenization patterns
// ============================================================

const WORD_RE = /\b[\w']+\b/g;
const SENTENCE_END_RE = /[.!?]+/;
// Emoji ranges: misc symbols + dingbats + emoji extensions + regional
// indicators. Same charset class as the Python version.
const EMOJI_RE =
  /[\u{1F300}-\u{1FAFF}\u{2600}-\u{27BF}\u{1F1E0}-\u{1F1FF}]/gu;
const EM_DASH_RE = /—|--|\s-\s/g;
const ELLIPSIS_RE = /\.{3,}|…/g;
const SEMICOLON_RE = /;/g;
const PARENTHETICAL_RE = /\([^)]{4,80}\)/g;
const NUMBER_RE = /\b\d+(?:[.,]\d+)?\b/g;

const TRACKED_FILLERS: readonly string[] = [
  'tbh',
  'ngl',
  'imo',
  'imho',
  'fwiw',
  'idk',
  'honestly',
  'literally',
  'actually',
  'anyway',
  'ok so',
  'alright',
  'look',
  'hot take',
  'real talk',
  'fr',
  'lowkey',
  'highkey',
];

const TRACKED_OPENERS: readonly string[] = [
  'ok so',
  'alright',
  'tbh',
  'honestly',
  'look',
  'hot take',
  'real talk',
  "i've been thinking",
  'quick thought',
  'thinking about',
];

const TRACKED_CLOSERS: readonly string[] = [
  'anyway',
  'idk',
  'fwiw',
  'edit:',
  'tldr:',
  'thoughts?',
  "we'll see",
  "here's hoping",
];

const HEDGE_MARKERS: readonly string[] = [
  'about',
  'around',
  'roughly',
  'approximately',
  'i think',
  'maybe',
  'give or take',
  'more or less',
  'ish',
  'kinda',
];

const SELF_CORRECTION_MARKERS: readonly string[] = [
  'actually wait',
  'scratch that',
  'no wait',
  'hmm actually',
  'i meant',
  'what i meant',
  'let me restart',
];

const TRACKED_PROFANITY: readonly string[] = [
  'shit',
  'fuck',
  'damn',
  'hell',
  'ass',
  'bullshit',
  'crap',
  'wtf',
  'tf',
  'af',
];

// ============================================================
// Per-pattern helpers
// ============================================================

function per1000Words(count: number, totalWords: number): number {
  if (totalWords === 0) return 0;
  return Math.round((count / totalWords) * 1000 * 100) / 100;
}

function countMatches(re: RegExp, text: string): number {
  // We rebuild the regex per call so the global flag state doesn't
  // leak between invocations.
  const fresh = new RegExp(re.source, re.flags);
  let n = 0;
  while (fresh.exec(text) !== null) n++;
  return n;
}

function wordsOf(text: string): string[] {
  return text.match(WORD_RE) ?? [];
}

// ============================================================
// Public API
// ============================================================

export const MIN_POSTS_FOR_EXTRACTION = 10;
const TRIM_PERCENT = 0.1; // Trim top/bottom 10% by length

export interface ExtractorInputPost {
  text: string;
}

/**
 * Run statistical analysis on a list of past posts and produce a
 * structured VoiceIdiosyncrasies object.
 *
 * Returns null when there aren't enough posts to extract reliable
 * patterns. Callers (the prompt builder) handle null by skipping
 * the WRITER VOICE PROFILE block — the model falls back to brand
 * bible + voice fingerprint samples for tone guidance.
 */
export function extractVoiceIdiosyncrasies(
  posts: readonly ExtractorInputPost[],
): VoiceIdiosyncrasies | null {
  if (posts.length < MIN_POSTS_FOR_EXTRACTION) return null;

  // Trim outliers by text length (top/bottom 10%).
  const byLength = [...posts].sort((a, b) => a.text.length - b.text.length);
  const trimN = Math.floor(byLength.length * TRIM_PERCENT);
  const trimmed = trimN > 0 ? byLength.slice(trimN, -trimN) : byLength;
  if (trimmed.length < MIN_POSTS_FOR_EXTRACTION) return null;

  const texts = trimmed.map((p) => p.text);
  const combined = texts.join('\n\n');
  const totalWords = wordsOf(combined).length;

  const punctuation = {
    emDashPer1000Words: per1000Words(countMatches(EM_DASH_RE, combined), totalWords),
    ellipsisPer1000Words: per1000Words(
      countMatches(ELLIPSIS_RE, combined),
      totalWords,
    ),
    semicolonPer1000Words: per1000Words(
      countMatches(SEMICOLON_RE, combined),
      totalWords,
    ),
    parentheticalAsidePer1000Words: per1000Words(
      countMatches(PARENTHETICAL_RE, combined),
      totalWords,
    ),
  };

  const lowercaseFirstLetterRatio = roundTo(
    texts.filter((t) => {
      const stripped = t.trim();
      const first = stripped[0];
      return first !== undefined && /^[a-z]$/.test(first);
    }).length / texts.length,
    2,
  );

  const commonFillerWords: Record<string, number> = {};
  for (const filler of TRACKED_FILLERS) {
    const postCount = texts.filter((t) => t.toLowerCase().includes(filler))
      .length;
    if (postCount > 0) {
      commonFillerWords[filler] = roundTo(postCount / texts.length, 2);
    }
  }

  const commonOpeners = extractCommonOpeners(texts);
  const commonClosers = extractCommonClosers(texts);

  const sentences = combined
    .split(SENTENCE_END_RE)
    .map((s) => s.trim())
    .filter(Boolean);
  let avgSentenceLengthWords = 0;
  let fragmentRatio = 0;
  if (sentences.length > 0) {
    const lengths = sentences.map((s) => wordsOf(s).length);
    avgSentenceLengthWords =
      Math.round(
        (lengths.reduce((a, b) => a + b, 0) / lengths.length) * 10,
      ) / 10;
    const fragmentCount = lengths.filter((l) => l <= 4).length;
    fragmentRatio = roundTo(fragmentCount / sentences.length, 2);
  }

  let allEmojis: string[] = [];
  for (const t of texts) {
    allEmojis = allEmojis.concat(t.match(EMOJI_RE) ?? []);
  }
  const emojiPerPost = roundTo(allEmojis.length / texts.length, 2);
  const emojiCounts = countBy(allEmojis);
  const commonEmojis = topN(emojiCounts, 5);

  const combinedLower = combined.toLowerCase();
  const usedProfanity: string[] = [];
  let totalProfanity = 0;
  for (const word of TRACKED_PROFANITY) {
    const wordRe = new RegExp(`\\b${escapeRegex(word)}\\b`, 'gi');
    const count = countMatches(wordRe, combinedLower);
    if (count > 0) {
      usedProfanity.push(word);
      totalProfanity += count;
    }
  }

  const hedgingRatio = computeHedgingRatio(combinedLower);
  const selfCorrectionCount = TRACKED_SELF_CORRECTION_COUNT(combinedLower);

  return {
    sampleSize: trimmed.length,
    extractedAt: new Date().toISOString(),
    ...punctuation,
    lowercaseFirstLetterRatio,
    commonFillerWords,
    commonProfanity: usedProfanity,
    profanityPer1000Words: per1000Words(totalProfanity, totalWords),
    avgSentenceLengthWords,
    fragmentRatio,
    emojiPerPost,
    commonEmojis,
    commonOpeners,
    commonClosers,
    hedgingRatio,
    selfCorrectionCount,
  };
}

// ============================================================
// Helpers
// ============================================================

function extractCommonOpeners(texts: readonly string[]): string[] {
  const counter = new Map<string, number>();
  for (const post of texts) {
    const head = post.toLowerCase().trim().slice(0, 30);
    for (const opener of TRACKED_OPENERS) {
      if (head.startsWith(opener)) {
        counter.set(opener, (counter.get(opener) ?? 0) + 1);
      }
    }
  }
  return topN(counter, 5);
}

function extractCommonClosers(texts: readonly string[]): string[] {
  const counter = new Map<string, number>();
  for (const post of texts) {
    const tail = post.toLowerCase().trim().slice(-50);
    for (const closer of TRACKED_CLOSERS) {
      if (tail.includes(closer)) {
        counter.set(closer, (counter.get(closer) ?? 0) + 1);
      }
    }
  }
  return topN(counter, 5);
}

function computeHedgingRatio(textLower: string): number {
  const fresh = new RegExp(NUMBER_RE.source, NUMBER_RE.flags);
  const matches: RegExpExecArray[] = [];
  let m: RegExpExecArray | null;
  while ((m = fresh.exec(textLower)) !== null) {
    matches.push(m);
  }
  if (matches.length === 0) return 0;

  let hedged = 0;
  for (const match of matches) {
    const windowStart = Math.max(0, match.index - 30);
    const window = textLower.slice(windowStart, match.index + match[0].length);
    if (HEDGE_MARKERS.some((marker) => window.includes(marker))) hedged++;
  }
  return roundTo(hedged / matches.length, 2);
}

function TRACKED_SELF_CORRECTION_COUNT(textLower: string): number {
  return SELF_CORRECTION_MARKERS.filter((m) => textLower.includes(m)).length;
}

function roundTo(value: number, digits: number): number {
  const factor = 10 ** digits;
  return Math.round(value * factor) / factor;
}

function countBy<T>(items: readonly T[]): Map<T, number> {
  const counter = new Map<T, number>();
  for (const item of items) {
    counter.set(item, (counter.get(item) ?? 0) + 1);
  }
  return counter;
}

function topN<T>(counter: Map<T, number>, n: number): T[] {
  return [...counter.entries()]
    .sort((a, b) => b[1] - a[1])
    .slice(0, n)
    .map(([k]) => k);
}

function escapeRegex(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

// ============================================================
// Format helper — produces the WRITER VOICE PROFILE prompt block
// ============================================================

export function formatIdiosyncrasiesAsPromptRules(
  idio: VoiceIdiosyncrasies,
): string {
  const lines: string[] = [];

  lines.push(
    `WRITER VOICE PROFILE (analyzed from last ${idio.sampleSize} posts):`,
  );
  lines.push('');
  lines.push('PUNCTUATION PATTERNS:');
  lines.push(
    `  - Em dashes: ${idio.emDashPer1000Words} per 1000 words (${describeFrequency(idio.emDashPer1000Words)})`,
  );
  lines.push(
    `  - Ellipsis: ${idio.ellipsisPer1000Words} per 1000 words (${describeFrequency(idio.ellipsisPer1000Words)})`,
  );
  lines.push(`  - Semicolons: ${idio.semicolonPer1000Words} per 1000 words`);
  lines.push(
    `  - Parenthetical asides: ${idio.parentheticalAsidePer1000Words} per 1000 words`,
  );
  lines.push('');
  lines.push('STRUCTURE:');
  lines.push(`  - Average sentence length: ${idio.avgSentenceLengthWords} words`);
  lines.push(
    `  - Fragment ratio: ${Math.round(idio.fragmentRatio * 100)}% of sentences are fragments`,
  );
  lines.push(
    `  - Lowercase first letter: ${Math.round(idio.lowercaseFirstLetterRatio * 100)}% of posts`,
  );
  lines.push('');
  lines.push('VOCABULARY:');

  const fillerEntries = Object.entries(idio.commonFillerWords)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 5);
  if (fillerEntries.length > 0) {
    lines.push('  - Filler words used:');
    for (const [word, freq] of fillerEntries) {
      lines.push(`    - '${word}': used in ${Math.round(freq * 100)}% of posts`);
    }
  } else {
    lines.push('  - No tracked filler words found.');
  }

  if (idio.commonProfanity.length > 0) {
    lines.push(
      `  - Profanity: ${idio.profanityPer1000Words} per 1000 words; common: ${idio.commonProfanity.join(', ')}`,
    );
  } else {
    lines.push('  - No profanity in tracked sample.');
  }

  lines.push('');
  lines.push('EMOJI:');
  lines.push(`  - ${idio.emojiPerPost} emojis per post on average`);
  if (idio.commonEmojis.length > 0) {
    lines.push(`  - Common emojis: ${idio.commonEmojis.join(' ')}`);
  }

  lines.push('');
  lines.push('OPENERS / CLOSERS:');
  if (idio.commonOpeners.length > 0) {
    lines.push(`  - Common openers: ${idio.commonOpeners.join(', ')}`);
  }
  if (idio.commonClosers.length > 0) {
    lines.push(`  - Common closers: ${idio.commonClosers.join(', ')}`);
  }

  lines.push('');
  lines.push('NUMBERS AND CORRECTIONS:');
  lines.push(
    `  - Number hedging: ${Math.round(idio.hedgingRatio * 100)}% of numbers are hedged ('about X', 'around Y'). Match this hedge ratio.`,
  );
  lines.push(
    `  - Self-correction frequency: ${idio.selfCorrectionCount} occurrences in sample.`,
  );

  lines.push('');
  lines.push('APPLICATION RULES:');
  lines.push(
    '  - Match these patterns approximately, not mechanically. If em dash usage',
  );
  lines.push(
    '    is 0.2 per 1000 words, use 0 in a 500-word post (matches the rate).',
  );
  lines.push('  - Filler words appear in % of posts, not every post. Vary use.');
  lines.push(
    '  - Lowercase first letter ratio is per post; use it as a probabilistic guide.',
  );

  return lines.join('\n');
}

function describeFrequency(per1000: number): string {
  if (per1000 === 0) return 'never used';
  if (per1000 < 0.5) return 'very rare';
  if (per1000 < 2) return 'occasional';
  if (per1000 < 5) return 'moderate';
  return 'frequent';
}

// ============================================================
// Staleness helper
// ============================================================

const STALE_DAYS = 7;

export function isIdiosyncrasiesStale(
  idio: VoiceIdiosyncrasies,
  asOf: Date = new Date(),
): boolean {
  const extractedAt = new Date(idio.extractedAt).getTime();
  const ageMs = asOf.getTime() - extractedAt;
  return ageMs > STALE_DAYS * 24 * 60 * 60 * 1000;
}

File: lib/voice-engine/voice_idiosyncrasy_extractor.py
"""
voice_idiosyncrasy_extractor.py
================================

Extracts structured voice idiosyncrasies from a client's accumulated posts.
The output is a VoiceIdiosyncrasies object that gets injected into the
generation prompt as concrete rules (not passive examples).

Recommended schedule: run nightly per (client, platform) once the client has
>= 10 posts on that platform. Re-run with sliding 30-day window to catch
voice drift.

Version: 1.0 (Phase 1.5)
"""

from __future__ import annotations

import re
import statistics
from collections import Counter
from datetime import datetime, timezone

from client_context import VoiceIdiosyncrasies, WeightedPost


# ============================================================================
# Tokenization helpers
# ============================================================================

WORD_RE = re.compile(r"\b[\w']+\b")
SENTENCE_END_RE = re.compile(r"[.!?]+")
EMOJI_RE = re.compile(
    "["
    "\U0001F300-\U0001FAFF"
    "\U00002600-\U000027BF"
    "\U0001F1E0-\U0001F1FF"
    "]"
)
EM_DASH_RE = re.compile(r"—|--|\s-\s")
ELLIPSIS_RE = re.compile(r"\.{3,}|…")
SEMICOLON_RE = re.compile(r";")
PARENTHETICAL_RE = re.compile(r"\([^)]{4,80}\)")  # Real asides, not citations

# Common filler words to track per-client preference
TRACKED_FILLERS: tuple[str, ...] = (
    "tbh", "ngl", "imo", "imho", "fwiw", "idk", "honestly",
    "literally", "actually", "anyway", "ok so", "alright", "look",
    "hot take", "real talk", "fr", "lowkey", "highkey",
)

# Common openers and closers
TRACKED_OPENERS: tuple[str, ...] = (
    "ok so", "alright", "tbh", "honestly", "look", "hot take",
    "real talk", "i've been thinking", "quick thought", "thinking about",
)

TRACKED_CLOSERS: tuple[str, ...] = (
    "anyway", "idk", "fwiw", "edit:", "tldr:", "thoughts?",
    "we'll see", "here's hoping",
)

# Hedging markers (vs precise numbers)
HEDGE_MARKERS: tuple[str, ...] = (
    "about", "around", "roughly", "approximately", "i think",
    "maybe", "give or take", "more or less", "ish", "kinda",
)

# Self-correction markers
SELF_CORRECTION_MARKERS: tuple[str, ...] = (
    "actually wait", "scratch that", "no wait", "hmm actually",
    "i meant", "what i meant", "let me restart",
)

# Profanity (mild list, expand per cultural context)
TRACKED_PROFANITY: tuple[str, ...] = (
    "shit", "fuck", "damn", "hell", "ass", "bullshit", "crap",
    "wtf", "tf", "af",
)


# ============================================================================
# Per-pattern extractors
# ============================================================================

def per_1000_words(count: int, total_words: int) -> float:
    """Normalize a count to a per-1000-words rate."""
    if total_words == 0:
        return 0.0
    return round((count / total_words) * 1000, 2)


def extract_punctuation_stats(combined_text: str, total_words: int) -> dict:
    return {
        "em_dash_per_1000_words": per_1000_words(
            len(EM_DASH_RE.findall(combined_text)), total_words
        ),
        "ellipsis_per_1000_words": per_1000_words(
            len(ELLIPSIS_RE.findall(combined_text)), total_words
        ),
        "semicolon_per_1000_words": per_1000_words(
            len(SEMICOLON_RE.findall(combined_text)), total_words
        ),
        "parenthetical_aside_per_1000_words": per_1000_words(
            len(PARENTHETICAL_RE.findall(combined_text)), total_words
        ),
    }


def extract_filler_word_frequencies(
    posts: list[str], total_posts: int
) -> dict[str, float]:
    """For each tracked filler, return % of posts that contain it at least once."""
    if total_posts == 0:
        return {}
    frequencies: dict[str, float] = {}
    for filler in TRACKED_FILLERS:
        post_count = sum(1 for p in posts if filler in p.lower())
        if post_count > 0:
            frequencies[filler] = round(post_count / total_posts, 2)
    return frequencies


def extract_opener_patterns(posts: list[str]) -> list[str]:
    """Return the openers the writer actually uses, ordered by frequency."""
    counter: Counter[str] = Counter()
    for post in posts:
        first_words = post.lower().strip()[:30]
        for opener in TRACKED_OPENERS:
            if first_words.startswith(opener):
                counter[opener] += 1
    return [opener for opener, _ in counter.most_common(5)]


def extract_closer_patterns(posts: list[str]) -> list[str]:
    counter: Counter[str] = Counter()
    for post in posts:
        last_words = post.lower().strip()[-50:]
        for closer in TRACKED_CLOSERS:
            if closer in last_words:
                counter[closer] += 1
    return [closer for closer, _ in counter.most_common(5)]


def extract_lowercase_first_letter_ratio(posts: list[str]) -> float:
    """Ratio of posts that start with a lowercase letter."""
    if not posts:
        return 0.0
    lowercase_starts = sum(
        1 for p in posts
        if p.strip() and p.strip()[0].islower()
    )
    return round(lowercase_starts / len(posts), 2)


def extract_sentence_stats(combined_text: str) -> dict:
    """Average sentence length and fragment ratio."""
    sentences = [s.strip() for s in SENTENCE_END_RE.split(combined_text) if s.strip()]
    if not sentences:
        return {"avg_sentence_length_words": 0, "fragment_ratio": 0.0}

    lengths = [len(WORD_RE.findall(s)) for s in sentences]
    avg_len = round(statistics.mean(lengths), 1)

    # Fragment heuristic: <= 4 words AND no verb suffix in common verbs
    fragment_count = sum(1 for length in lengths if length <= 4)
    fragment_ratio = round(fragment_count / len(sentences), 2)

    return {
        "avg_sentence_length_words": avg_len,
        "fragment_ratio": fragment_ratio,
    }


def extract_emoji_patterns(posts: list[str], total_posts: int) -> dict:
    if total_posts == 0:
        return {"emoji_per_post": 0.0, "common_emojis": []}
    all_emojis: list[str] = []
    for post in posts:
        all_emojis.extend(EMOJI_RE.findall(post))
    counter = Counter(all_emojis)
    return {
        "emoji_per_post": round(len(all_emojis) / total_posts, 2),
        "common_emojis": [emoji for emoji, _ in counter.most_common(5)],
    }


def extract_profanity_stats(combined_text: str, total_words: int) -> dict:
    text_lower = combined_text.lower()
    used: list[str] = []
    total_count = 0
    for word in TRACKED_PROFANITY:
        count = len(re.findall(rf"\b{re.escape(word)}\b", text_lower))
        if count > 0:
            used.append(word)
            total_count += count
    return {
        "profanity_per_1000_words": per_1000_words(total_count, total_words),
        "common_profanity": used,
    }


def extract_hedging_ratio(combined_text: str) -> float:
    """Ratio of numerical references that are hedged.

    Counts: number of numbers preceded or followed (within 3 words) by a
    hedge marker, divided by total numbers. Returns 0-1.
    """
    text_lower = combined_text.lower()
    numbers = list(re.finditer(r"\b\d+(?:[.,]\d+)?\b", text_lower))
    if not numbers:
        return 0.0

    hedged_count = 0
    for match in numbers:
        # Look 30 chars before for hedge markers
        window_start = max(0, match.start() - 30)
        window = text_lower[window_start : match.end()]
        if any(marker in window for marker in HEDGE_MARKERS):
            hedged_count += 1
    return round(hedged_count / len(numbers), 2)


def extract_self_correction_count(combined_text: str) -> int:
    text_lower = combined_text.lower()
    return sum(1 for marker in SELF_CORRECTION_MARKERS if marker in text_lower)


# ============================================================================
# Public API
# ============================================================================

MIN_POSTS_FOR_EXTRACTION = 10
TRIM_PERCENT = 0.10  # Trim top/bottom 10% by length to remove outliers


def extract_voice_idiosyncrasies(
    posts: list[WeightedPost],
    *,
    min_posts: int = MIN_POSTS_FOR_EXTRACTION,
) -> VoiceIdiosyncrasies | None:
    """Run statistical analysis on a list of past posts and produce
    structured voice idiosyncrasies.

    Returns None if there aren't enough posts to extract reliable patterns.
    Caller should fall back to baseline behavior in that case.
    """
    if len(posts) < min_posts:
        return None

    # Trim outliers by post length
    posts_by_length = sorted(posts, key=lambda p: len(p.text))
    trim_n = int(len(posts_by_length) * TRIM_PERCENT)
    if trim_n > 0:
        trimmed = posts_by_length[trim_n:-trim_n]
    else:
        trimmed = posts_by_length

    if len(trimmed) < min_posts:
        return None

    texts = [p.text for p in trimmed]
    combined_text = "\n\n".join(texts)
    total_words = len(WORD_RE.findall(combined_text))

    punctuation = extract_punctuation_stats(combined_text, total_words)
    sentence_stats = extract_sentence_stats(combined_text)
    emoji_stats = extract_emoji_patterns(texts, len(trimmed))
    profanity_stats = extract_profanity_stats(combined_text, total_words)

    return VoiceIdiosyncrasies(
        sample_size=len(trimmed),
        last_extracted=datetime.now(timezone.utc),
        em_dash_per_1000_words=punctuation["em_dash_per_1000_words"],
        ellipsis_per_1000_words=punctuation["ellipsis_per_1000_words"],
        semicolon_per_1000_words=punctuation["semicolon_per_1000_words"],
        parenthetical_aside_per_1000_words=punctuation["parenthetical_aside_per_1000_words"],
        lowercase_first_letter_ratio=extract_lowercase_first_letter_ratio(texts),
        common_filler_words=extract_filler_word_frequencies(texts, len(trimmed)),
        avg_sentence_length_words=sentence_stats["avg_sentence_length_words"],
        fragment_ratio=sentence_stats["fragment_ratio"],
        profanity_per_1000_words=profanity_stats["profanity_per_1000_words"],
        common_profanity=profanity_stats["common_profanity"],
        emoji_per_post=emoji_stats["emoji_per_post"],
        common_emojis=emoji_stats["common_emojis"],
        common_openers=extract_opener_patterns(texts),
        common_closers=extract_closer_patterns(texts),
        hedging_ratio=extract_hedging_ratio(combined_text),
        self_correction_count=extract_self_correction_count(combined_text),
    )


def format_idiosyncrasies_as_prompt_rules(idio: VoiceIdiosyncrasies) -> str:
    """Format a VoiceIdiosyncrasies object as concrete rules text for the
    generation prompt. The model receives this as the WRITER VOICE PROFILE
    section and applies these patterns rather than passively imitating examples.
    """
    lines = [
        f"WRITER VOICE PROFILE (analyzed from last {idio.sample_size} posts):",
        "",
        "PUNCTUATION PATTERNS:",
        f"  - Em dashes: {idio.em_dash_per_1000_words} per 1000 words "
        f"({_describe_frequency(idio.em_dash_per_1000_words, 'em_dash')})",
        f"  - Ellipsis: {idio.ellipsis_per_1000_words} per 1000 words "
        f"({_describe_frequency(idio.ellipsis_per_1000_words, 'ellipsis')})",
        f"  - Semicolons: {idio.semicolon_per_1000_words} per 1000 words",
        f"  - Parenthetical asides: {idio.parenthetical_aside_per_1000_words} per 1000 words",
        "",
        "STRUCTURE:",
        f"  - Average sentence length: {idio.avg_sentence_length_words} words",
        f"  - Fragment ratio: {int(idio.fragment_ratio * 100)}% of sentences are fragments",
        f"  - Lowercase first letter: {int(idio.lowercase_first_letter_ratio * 100)}% of posts",
        "",
        "VOCABULARY:",
    ]

    if idio.common_filler_words:
        filler_lines = [
            f"    - '{word}': used in {int(freq * 100)}% of posts"
            for word, freq in sorted(idio.common_filler_words.items(),
                                     key=lambda x: x[1], reverse=True)[:5]
        ]
        lines.append("  - Filler words used:")
        lines.extend(filler_lines)
    else:
        lines.append("  - No tracked filler words found.")

    if idio.common_profanity:
        lines.append(
            f"  - Profanity: {idio.profanity_per_1000_words} per 1000 words; "
            f"common: {', '.join(idio.common_profanity)}"
        )
    else:
        lines.append("  - No profanity in tracked sample.")

    lines.extend([
        "",
        "EMOJI:",
        f"  - {idio.emoji_per_post} emojis per post on average",
    ])
    if idio.common_emojis:
        lines.append(f"  - Common emojis: {' '.join(idio.common_emojis)}")

    lines.extend([
        "",
        "OPENERS / CLOSERS:",
    ])
    if idio.common_openers:
        lines.append(f"  - Common openers: {', '.join(idio.common_openers)}")
    if idio.common_closers:
        lines.append(f"  - Common closers: {', '.join(idio.common_closers)}")

    lines.extend([
        "",
        "NUMBERS AND CORRECTIONS:",
        f"  - Number hedging: {int(idio.hedging_ratio * 100)}% of numbers are hedged "
        f"('about X', 'around Y'). Match this hedge ratio.",
        f"  - Self-correction frequency: {idio.self_correction_count} occurrences in sample.",
        "",
        "APPLICATION RULES:",
        "  - Match these patterns approximately, not mechanically. If em dash usage",
        "    is 0.2 per 1000 words, use 0 in a 500-word post (matches the rate).",
        "  - Filler words appear in % of posts, not every post. Vary use.",
        "  - Lowercase first letter ratio is per post; use it as a probabilistic guide.",
    ])

    return "\n".join(lines)


def _describe_frequency(per_1000: float, kind: str) -> str:
    """Human-readable frequency description."""
    if per_1000 == 0:
        return "never used"
    if per_1000 < 0.5:
        return "very rare"
    if per_1000 < 2:
        return "occasional"
    if per_1000 < 5:
        return "moderate"
    return "frequent"
